Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in

```diff
@@ -50,4 +50,71 @@
  * Marks a symbol as deprecated, indicating a replacement.
  */
 
+#if ENABLE(2022_GLIB_API)
+/*
+ * The G_DECLARE_DERIVABLE_TYPE macro creates an instance struct that
+ * has ParentName as its only member; but in order to keep the same code
+ * for the pre-2022 API it should have also a "priv" member. There is no
+ * reasonable way of bending the GLib macro to do what we want; so this is
+ * a version of the GLib macro adapted accordingly.
+ */
+#define WEBKIT_DECLARE_TYPE(ModuleObjName, module_obj_name, MODULE, OBJ_NAME, ParentName)                \
+  WEBKIT_API GType module_obj_name##_get_type (void);                                                    \
+  G_GNUC_BEGIN_IGNORE_DEPRECATIONS                                                                       \
+  typedef struct _##ModuleObjName ModuleObjName;                                                         \
+  typedef struct _##ModuleObjName##Class ModuleObjName##Class;                                           \
+  typedef struct _##ModuleObjName##Private ModuleObjName##Private;                                       \
+  struct _##ModuleObjName { ParentName parent_instance; ModuleObjName ## Private *priv; };               \
+                                                                                                         \
+  _GLIB_DEFINE_AUTOPTR_CHAINUP (ModuleObjName, ParentName)                                               \
+  G_DEFINE_AUTOPTR_CLEANUP_FUNC (ModuleObjName##Class, g_type_class_unref)                               \
+                                                                                                         \
+  G_GNUC_UNUSED static inline ModuleObjName * MODULE##_##OBJ_NAME (gpointer ptr) {                       \
+    return G_TYPE_CHECK_INSTANCE_CAST (ptr, module_obj_name##_get_type (), ModuleObjName); }             \
+  G_GNUC_UNUSED static inline ModuleObjName##Class * MODULE##_##OBJ_NAME##_CLASS (gpointer ptr) {        \
+    return G_TYPE_CHECK_CLASS_CAST (ptr, module_obj_name##_get_type (), ModuleObjName##Class); }         \
+  G_GNUC_UNUSED static inline gboolean MODULE##_IS_##OBJ_NAME (gpointer ptr) {                           \
+    return G_TYPE_CHECK_INSTANCE_TYPE (ptr, module_obj_name##_get_type ()); }                            \
+  G_GNUC_UNUSED static inline gboolean MODULE##_IS_##OBJ_NAME##_CLASS (gpointer ptr) {                   \
+    return G_TYPE_CHECK_CLASS_TYPE (ptr, module_obj_name##_get_type ()); }                               \
+  G_GNUC_UNUSED static inline ModuleObjName##Class * MODULE##_##OBJ_NAME##_GET_CLASS (gpointer ptr) {    \
+    return G_TYPE_INSTANCE_GET_CLASS (ptr, module_obj_name##_get_type (), ModuleObjName##Class); }       \
+  G_GNUC_END_IGNORE_DEPRECATIONS
+
+#define WEBKIT_DECLARE_DERIVABLE_TYPE WEBKIT_DECLARE_TYPE
```

Comment:
Nit: the `WEBKIT_DECLARE_TYPE` is unneeded because it's only used here. Can simply rename it to `WEBKIT_DECLARE_DERIVABLE_TYPE`.

BTW, I agree this is the best-possible solution.

Reply:
The `WEBKIT_DECLARE_TYPE` macro is used in `WebKitDownload.h.in`, `WebKitWebContext.h.in`, and `WebKitWebHitTestResult.h.in`. While right now it's the same as `WEBKIT_DECLARE_DERIVABLE_TYPE`, I would find it confusing that we use something named “derivable” to define a final type like `WebKitWebContext` or `WebKitDownload`, and at the same time we cannot use `WEBKIT_DECLARE_FINAL_TYPE` there because it hardcodes the class struct to contain nothing but these types do use a custom class struct with vtable functions.

Reply:
OK... but the custom class struct does not belong in the modern API version because the type is no longer derivable and so there's no point to having vfuncs because they can never be used. So I would use `WEBKIT_DECLARE_FINAL_TYPE` anyway and move the class struct inside `#if !ENABLE(2022_GLIB_API)`.

Beware that does require some small changes in the implementation file, e.g. for WebKitDownload:

```
diff --git a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
index 44367ee66541..b3459268c1cd 100644
--- a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
+++ b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
@@ -283,7 +283,11 @@ static void webkit_download_class_init(WebKitDownloadClass* downloadClass)
         "decide-destination",
         G_TYPE_FROM_CLASS(objectClass),
         G_SIGNAL_RUN_LAST,
+#if ENABLE(2022_GLIB_API)
+        0,
+#else
         G_STRUCT_OFFSET(WebKitDownloadClass, decide_destination),
+#endif
         g_signal_accumulator_true_handled, NULL,
         g_cclosure_marshal_generic,
         G_TYPE_BOOLEAN, 1,
```

Reply:
@mcatanzaro It's not that easy, the above diff is not enough, what you want is this:

```diff
diff --git a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
index 586a01a90ea9..c7c4cdf52c4e 100644
--- a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
+++ b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
@@ -150,7 +150,9 @@ static void webkit_download_class_init(WebKitDownloadClass* downloadClass)
     objectClass->set_property = webkitDownloadSetProperty;
     objectClass->get_property = webkitDownloadGetProperty;
 
+#if !ENABLE(2022_GLIB_API)
     downloadClass->decide_destination = webkitDownloadDecideDestination;
+#endif
 
     /**
      * WebKitDownload:destination:
@@ -283,7 +285,11 @@ static void webkit_download_class_init(WebKitDownloadClass* downloadClass)
         "decide-destination",
         G_TYPE_FROM_CLASS(objectClass),
         G_SIGNAL_RUN_LAST,
-        G_STRUCT_OFFSET(WebKitDownloadClass, decide_destination),
+#if ENABLE(2022_GLIB_API)
+        0,
+#else
+        G_STRUCT_OFFSET(WebKitDownloadClass,  decide_destination),
+#endif
         g_signal_accumulator_true_handled, NULL,
         g_cclosure_marshal_generic,
         G_TYPE_BOOLEAN, 1,
@@ -408,8 +414,16 @@ String webkitDownloadDecideDestinationWithSuggestedFilename(WebKitDownload* down
 {
     if (download->priv->isCancelled)
         return emptyString();
-    gboolean returnValue;
-    g_signal_emit(download, signals[DECIDE_DESTINATION], 0, suggestedFilename.data(), &returnValue);
+
+    gboolean handled = FALSE;
+    g_signal_emit(download, signals[DECIDE_DESTINATION], 0, suggestedFilename.data(), &handled);
+
+#if ENABLE(2022_GLIB_API)
+    if (!handled)
+        handled = webkitDownloadDecideDestination(download, suggestedFilename.data());
+#endif
+    ASSERT(handled);
+
     allowOverwrite = download->priv->allowOverwrite;
     GUniquePtr<char> destinationPath(g_filename_from_uri(download->priv->destinationURI.data(), nullptr, nullptr));
     if (!destinationPath)
```

While not too complicated, and in most cases (if not all) the other types which currently have a public class struct with vfuncs can be fixed in similar ways&mdash;but this patch is already big enough, so I would rather do these changes in a follow-up patch. WDYT?

Reply:
Sure, of course a follow-up patch is fine.

Reply:
So... r+ for this one, then? I think all the other issues have been tackled 😄